org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XSISchemaModel.java

```diff
@@ -0,0 +1,89 @@
+ *  http://www.eclipse.org/legal/epl-v10.html
+ *
+ *  Contributors:
+ *  Red Hat Inc. - initial API and implementation
+ */
+package org.eclipse.lsp4xml.services;
+
+import org.eclipse.lsp4j.CompletionItem;
+import org.eclipse.lsp4j.CompletionItemKind;
+import org.eclipse.lsp4j.InsertTextFormat;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.TextEdit;
+import org.eclipse.lsp4xml.dom.Element;
+import org.eclipse.lsp4xml.dom.XMLDocument;
+import org.eclipse.lsp4xml.services.extensions.ICompletionRequest;
+import org.eclipse.lsp4xml.services.extensions.ICompletionResponse;
+
+/**
+ * This class holds values that represent the XSI xsd.
+ * Can be seen at https://www.w3.org/2001/XMLSchema-instance
+ */
+public class XSISchemaModel {
+	
+	public static void computeCompletionResponses(ICompletionRequest request, 
+			ICompletionResponse response, Range editRange, XMLDocument document) {
+		String snippet = "";
+		boolean isSnippetsSupported = request.getCompletionSettings().isCompletionSnippetsSupported();
+		if(isSnippetsSupported) {
+			snippet = "$0";
+		}
+		String actualPrefix = document.getSchemaInstancePrefix();
+		String name;
+		String documentation;
+		Element root = document.getDocumentElement();
+		boolean schemaLocationExists = document.hasSchemaLocation();
+		boolean noNamespaceSchemaLocationExists = document.hasNoNamespaceSchemaLocation();
+		//Indicates that no values are allowed inside an XML element
+		if(!attributeAlreadyExists(root, actualPrefix, "nil")) {
+			documentation = "Indicates that no values are allowed inside an XML element";
+			name = actualPrefix + ":nil";
+			String nilSnippet = isSnippetsSupported ? "${0:true}" : "true";
+			createCompletionItem(name, isSnippetsSupported, nilSnippet, editRange, response, documentation);
+		}
+		//Signals that an element should be accepted as ·valid· when it has no content despite 
+		//a content type which does not require or even necessarily allow empty content. 
+		//An element may be ·valid· without content if it has the attribute xsi:nil with 
+		//the value true.
+		if(!attributeAlreadyExists(root, actualPrefix, "type")) {
+			documentation = "Signals that an element should be accepted as `valid` when it has no content despite " + 
+			"a content type which does not require or even necessarily allow empty content."; 
+			name = actualPrefix + ":type";
+			createCompletionItem(name, isSnippetsSupported, snippet, editRange, response, documentation);
+		}
+		//The xsi:schemaLocation and xsi:noNamespaceSchemaLocation attributes can be used in a document 
+		//to provide hints as to the physical location of schema documents which may be used for ·assessment·.
+		if(!schemaLocationExists && !noNamespaceSchemaLocationExists) {
+			documentation = "The xsi:schemaLocation attribute can be used in a document " +
+			"to provide hints as to the physical location of schema documents which may be used for ·assessment·.";
```

Comment:
extra space between for and assessment

Reply:
actually "references an XML Schema document that does have a target namespace" is better (from https://msdn.microsoft.com/en-us/library/ms256100(v=vs.110).aspx)